uber/templates/jobs/form.html

```diff
@@ -23,146 +40,168 @@
+        {{ options(c.DURATION_OPTS, job.duration) }}
+      </select>
+    </div>
+  </div>
+  {% if job.type == c.REGULAR %}
+    <div class="form-group">
+      <label class="col-sm-3 control-label">Extra 15{{ macros.popup_link("../static_views/extra15.html") }}</label>
+      <div class="col-sm-6">
+        <div class="form-control-static checkbox">
+          {{ macros.checkbox(job, 'extra15', label='Yes') }}
+        </div>
+      </div>
+    </div>
+  {% endif %}
+  <div class="form-group">
+    <label class="col-sm-3 control-label">Slots</label>
+    <div class="col-sm-6">
+      <select name="slots" class="form-control">
+        {{ int_options(1, 50, job.slots) }}
+      </select>
+      {% if not job.is_new %}<i>{{ job.shifts|length }} already filled</i>{% endif %}
+    </div>
+  </div>
+  {% if job.type == c.REGULAR %}
+    <div class="form-group">
+      <label class="col-sm-3 control-label">Weight{{ macros.popup_link("../static_views/weightDesc.html") }}</label>
+      <div class="col-sm-6">
+        <select name="weight" class="form-control">
+          {{ options(c.WEIGHT_OPTS, job.weight) }}
         </select>
-        {% if not job.is_new %}({{ job.shifts|length }} filled){% endif %}
-    </td>
-</tr>
-<tr>
-    <td> <b>Location:</b> </td>
-    <td>
-        <select name="location">
-            {{ options(c.JOB_LOCATION_OPTS,job.location) }}
-        </select>
-    </td>
-</tr>
-{% if job.type == c.REGULAR %}
-    <tr>
-        <td> <b>Weight:{{ macros.popup_link("../static_views/weightDesc.html") }}</b> </td>
-        <td>
-            <select name="weight">
-                {{ options(c.WEIGHT_OPTS,job.weight) }}
-            </select>
-        </td>
-    </tr>
-{% endif %}
-<tr>
-    <td> <b>Restricted:{{ macros.popup_link("../static_views/trustDesc.html") }}</b> </td>
-    <td class="checkbox"> {{ macros.checkbox(job, 'restricted', label='Yes') }} </td>
-</tr>
-<tr>
-    <td></td>
-    <td> <input type="submit" value="Upload" />
-</tr>
-</table>
-
+      </div>
+    </div>
+  {% endif %}
+  <div class="form-group">
+    <label class="col-sm-3 control-label">Department</label>
+    <div class="col-sm-6">
+      <select name="department_id" class="form-control">
+        {{ options(c.DEPARTMENT_OPTS, job.department_id) }}
+      </select>
+    </div>
+  </div>
+  <div class="form-group">
+    <label class="col-sm-3 control-label">Required Roles{{ macros.popup_link("../static_views/trustDesc.html") }}</label>
+    <div class="col-sm-9">
+      <div class="form-control-static">
+        {%- for department_id, dept_roles_opts in dept_roles.items() -%}
+          <div id="dept_roles-{{ department_id }}" class="dept_roles">
+            {{ macros.checkgroup_opts(
+                  'required_roles_ids',
+                  dept_roles_opts,
+                  defaults=job.required_roles_ids,
+                  include_empty_hidden=True) }}
+          </div>
+        {%- endfor -%}
+      </div>
```

Comment:
Nice.  This was one of the files I went into this review specifically looking for, to see how you handled the user switching between which department is selected in the dropdown.  I like this approach!

Reply:
Ultimately this approach is not the best. We really should be doing an ajax query when we select a department. I took the easy approach, because I figured in the short term we'll have at most 50 departments with 4 or 5 roles at most. This implementation is not sustainable in the long term though.

Almost all of our tables need to be reworked to support dynamic paging. It's a pervasive problem that we've been able to deal with because of our relatively small size.